teamcity-rest-client-impl/src/main/kotlin/org/jetbrains/teamcity/rest/coroutines/nodePinning.kt

```diff
@@ -0,0 +1,55 @@
+package org.jetbrains.teamcity.rest.coroutines
+
+import okhttp3.Cookie
+import okhttp3.CookieJar
+import okhttp3.HttpUrl
+import org.jetbrains.teamcity.rest.NodeSelector
+
+fun NodeSelector.toCookieJar(): CookieJar {
+    return when(this) {
+        is NodeSelector.PinToNode -> PinToNodeCookieJar(this.nodeId)
+        is NodeSelector.ServerControlled -> ServerControlledRouting()
+        is NodeSelector.Unspecified -> CookieJar.NO_COOKIES
+    }
+}
+
+private class ServerControlledRouting: CookieJar {
+    private var serverProvidedNodeId: String? = null
```

Comment:
I'm not sure, but probably it's worth marking this field `kotlin.jvm.Volatile` to be sure updated value will be seen in all threads in concurrent context

```kotlin
@Volatile
private var serverProvidedNodeId: String? = null
```